### DISCUSSION TITLE
2021-12-03 Core Tools Update

### DISCUSSION THREAD

**Discussion by austinabell** [near/community#142] [Developer Tools]
### near-sdk-rs
- Added a safe API for newly added RIPEMD-160 hash function (https://github.com/near/near-sdk-rs/pull/648)
- Implemented new iteration of `LookupSet` collection type under `near_sdk::store` (https://github.com/near/near-sdk-rs/pull/654)
- experimental wip: caching static `env` variables (https://github.com/near/near-sdk-rs/pull/655)
- WIP on new `TreeMap` caching implementation

### workspaces-rs
- Refactor library to not expose external types and updated client libraries (https://github.com/near/workspaces-rs/pull/25)

### nearcore
 - Exhausively document `near-account-id` (https://github.com/near/nearcore/pull/5575)
 - Clean-up the `near-account-id` API in preparedness for stabilization (https://github.com/near/nearcore/pull/5581)
 - Synced with @matklad on refactoring `near-jsonrpc-primitives`, decoupling its types from nearcore types (https://github.com/near/nearcore/issues/5516#issuecomment-981698916)

**Comment by maxzaver** [comment]
@austinabell . Is release of near-sdk-rs 4.0.0 blocked on the protocol change that allows spreading gas evenly among the receipts?

**Reply by austinabell** [reply]
> @austinabell . Is release of near-sdk-rs 4.0.0 blocked on the protocol change that allows spreading gas evenly among the receipts?

Given the long timeline of this protocol change is long and the API breakage is big, I think it's best to have this come in with the next version of the SDK, which can be developed in parallel with 4.0. 

The stabilization of 4.0 was blocked with the update of nearcore crates used for testing (https://github.com/near/near-sdk-rs/pull/639), but now there is nothing hard blocking the stabilization. There was one other thing on my tracking issue (https://github.com/near/near-sdk-rs/issues/494) which is https://github.com/near/near-sdk-rs/issues/263, but I don't see what else I can do with this to reach a consensus of the right direction.
